File: frameworks/nist.js
import { SecurityFramework, SecurityDomain, SecurityControl } from '../src/core/models.js';

/**
 * NIST Cybersecurity Framework Definition
 */
export function createNISTFramework() {
  const framework = new SecurityFramework(
    'nist-csf',
    'NIST Cybersecurity Framework',
    '1.1',
    'Framework for Improving Critical Infrastructure Cybersecurity'
  );

  // NIST CSF Core Functions
  const domains = [
    new SecurityDomain('ID', 'Identify', 'Develop an organizational understanding to manage cybersecurity risk', 'nist-csf'),
    new SecurityDomain('PR', 'Protect', 'Develop and implement appropriate safeguards to ensure delivery of critical services', 'nist-csf'),
    new SecurityDomain('DE', 'Detect', 'Develop and implement appropriate activities to identify the occurrence of a cybersecurity event', 'nist-csf'),
    new SecurityDomain('RS', 'Respond', 'Develop and implement appropriate activities to take action regarding a detected cybersecurity incident', 'nist-csf'),
    new SecurityDomain('RC', 'Recover', 'Develop and implement appropriate activities to maintain plans for resilience and to restore any capabilities or services', 'nist-csf')
  ];

  domains.forEach(domain => framework.addDomain(domain));

  // NIST CSF Core Controls
  const controls = [
    // Identify
    new SecurityControl(
      'ID.AM-1',
      'Asset Management - Physical Devices and Systems',
      'Physical devices and systems within the organization are inventoried',
      'nist-csf',
      'ID'
    ),
    new SecurityControl(
      'ID.AM-2',
      'Asset Management - Software Platforms and Applications',
      'Software platforms and applications within the organization are inventoried',
      'nist-csf',
      'ID'
    ),
    new SecurityControl(
      'ID.GV-1',
      'Governance - Information Security Policy',
      'Organizational cybersecurity policy is established and communicated',
      'nist-csf',
      'ID'
    ),
    new SecurityControl(
      'ID.RA-1',
      'Risk Assessment - Risk Management Process',
      'Asset vulnerabilities are identified and documented',
      'nist-csf',
      'ID'
    ),

    // Protect
    new SecurityControl(
      'PR.AC-1',
      'Access Control - Identity Management',
      'Identities and credentials are issued, managed, verified, revoked, and audited for authorized devices, users and processes',
      'nist-csf',
      'PR'
    ),
    new SecurityControl(
      'PR.AC-3',
      'Access Control - Remote Access',
      'Remote access is managed',
      'nist-csf',
      'PR'
    ),
    new SecurityControl(
      'PR.DS-1',
      'Data Security - Data-at-rest Protection',
      'Data-at-rest is protected',
      'nist-csf',
      'PR'
    ),
    new SecurityControl(
      'PR.DS-2',
      'Data Security - Data-in-transit Protection',
      'Data-in-transit is protected',
      'nist-csf',
      'PR'
    ),
    new SecurityControl(
      'PR.PT-1',
      'Protective Technology - Audit Logs',
      'Audit/log records are determined, documented, implemented, and reviewed',
      'nist-csf',
      'PR'
    ),

    // Detect
    new SecurityControl(
      'DE.AE-1',
      'Anomalies and Events - Baseline Establishment',
      'A baseline of network operations and expected data flows for users and systems is established and managed',
      'nist-csf',
      'DE'
    ),
    new SecurityControl(
      'DE.CM-1',
      'Security Continuous Monitoring - System Monitoring',
      'The network and physical environment is monitored to detect potential cybersecurity events',
      'nist-csf',
      'DE'
    ),

    // Respond
    new SecurityControl(
      'RS.RP-1',
      'Response Planning - Response Plan',
      'Response plan is executed during or after an incident',
      'nist-csf',
      'RS'
    ),
    new SecurityControl(
      'RS.CO-2',
      'Communications - Incident Reporting',
      'Incidents are reported consistent with established criteria',
      'nist-csf',
      'RS'
    ),

    // Recover
    new SecurityControl(
      'RC.RP-1',
      'Recovery Planning - Recovery Plan',
      'Recovery plan is executed during or after a cybersecurity incident',
      'nist-csf',
      'RC'
    ),
    new SecurityControl(
      'RC.IM-1',
      'Improvements - Lessons Learned',
      'Recovery plans incorporate lessons learned',
      'nist-csf',
      'RC'
    )
  ];

  controls.forEach(control => {
    control.addTag('nist');
    control.addTag('cybersecurity');
    control.addTag('framework');
    
    // Set control types
    if (control.id.includes('PR.AC') || control.id.includes('PR.DS') || control.id.includes('DE.CM')) {
      control.controlType = 'technical';
    } else if (control.id.includes('PR.PT')) {
      control.controlType = 'technical';
    }
    
    framework.addControl(control);
  });

  return framework;
}

File: mappings/frameworks/nist.py
"""
NIST Cybersecurity Framework Definition
"""

from ..core.models import SecurityFramework, SecurityDomain, SecurityControl, RiskLevel, ControlType


def create_nist_framework() -> SecurityFramework:
    """Create NIST Cybersecurity Framework definition."""
    framework = SecurityFramework(
        id="nist-csf",
        name="NIST Cybersecurity Framework",
        version="1.1",
        description="Framework for Improving Critical Infrastructure Cybersecurity"
    )

    # NIST CSF Core Functions
    domains = [
        SecurityDomain(
            id="ID",
            name="Identify",
            description="Develop an organizational understanding to manage cybersecurity risk",
            framework_id="nist-csf"
        ),
        SecurityDomain(
            id="PR",
            name="Protect",
            description="Develop and implement appropriate safeguards to ensure delivery of critical services",
            framework_id="nist-csf"
        ),
        SecurityDomain(
            id="DE",
            name="Detect",
            description="Develop and implement appropriate activities to identify the occurrence of a cybersecurity event",
            framework_id="nist-csf"
        ),
        SecurityDomain(
            id="RS",
            name="Respond",
            description="Develop and implement appropriate activities to take action regarding a detected cybersecurity incident",
            framework_id="nist-csf"
        ),
        SecurityDomain(
            id="RC",
            name="Recover",
            description="Develop and implement appropriate activities to maintain plans for resilience and to restore any capabilities or services",
            framework_id="nist-csf"
        )
    ]

    for domain in domains:
        framework.add_domain(domain)

    # NIST CSF Core Controls
    controls = [
        # Identify
        SecurityControl(
            id="ID.AM-1",
            title="Asset Management - Physical Devices and Systems",
            description="Physical devices and systems within the organization are inventoried",
            framework_id="nist-csf",
            domain_id="ID",
            risk_level=RiskLevel.HIGH,
            control_type=ControlType.PROCEDURAL,
            tags={"nist", "cybersecurity", "framework", "assets"}
        ),
        SecurityControl(
            id="ID.AM-2",
            title="Asset Management - Software Platforms and Applications",
            description="Software platforms and applications within the organization are inventoried",
            framework_id="nist-csf",
            domain_id="ID",
            risk_level=RiskLevel.HIGH,
            control_type=ControlType.PROCEDURAL,
            tags={"nist", "cybersecurity", "framework", "assets"}
        ),
        SecurityControl(
            id="ID.GV-1",
            title="Governance - Information Security Policy",
            description="Organizational cybersecurity policy is established and communicated",
            framework_id="nist-csf",
            domain_id="ID",
            risk_level=RiskLevel.CRITICAL,
            control_type=ControlType.PROCEDURAL,
            tags={"nist", "cybersecurity", "framework", "governance"}
        ),
        SecurityControl(
            id="ID.RA-1",
            title="Risk Assessment - Risk Management Process",
            description="Asset vulnerabilities are identified and documented",
            framework_id="nist-csf",
            domain_id="ID",
            risk_level=RiskLevel.HIGH,
            control_type=ControlType.PROCEDURAL,
            tags={"nist", "cybersecurity", "framework", "risk-assessment"}
        ),
        # Protect
        SecurityControl(
            id="PR.AC-1",
            title="Access Control - Identity Management",
            description="Identities and credentials are issued, managed, verified, revoked, and audited for authorized devices, users and processes",
            framework_id="nist-csf",
            domain_id="PR",
            risk_level=RiskLevel.CRITICAL,
            control_type=ControlType.TECHNICAL,
            tags={"nist", "cybersecurity", "framework", "access-control"}
        ),
        SecurityControl(
            id="PR.AC-3",
            title="Access Control - Remote Access",
            description="Remote access is managed",
            framework_id="nist-csf",
            domain_id="PR",
            risk_level=RiskLevel.HIGH,
            control_type=ControlType.TECHNICAL,
            tags={"nist", "cybersecurity", "framework", "access-control"}
        ),
        SecurityControl(
            id="PR.DS-1",
            title="Data Security - Data-at-rest Protection",
            description="Data-at-rest is protected",
            framework_id="nist-csf",
            domain_id="PR",
            risk_level=RiskLevel.HIGH,
            control_type=ControlType.TECHNICAL,
            tags={"nist", "cybersecurity", "framework", "data-security"}
        ),
        SecurityControl(
            id="PR.DS-2",
            title="Data Security - Data-in-transit Protection",
            description="Data-in-transit is protected",
            framework_id="nist-csf",
            domain_id="PR",
            risk_level=RiskLevel.HIGH,
            control_type=ControlType.TECHNICAL,
            tags={"nist", "cybersecurity", "framework", "data-security"}
        ),
        SecurityControl(
            id="PR.PT-1",
            title="Protective Technology - Audit Logs",
            description="Audit/log records are determined, documented, implemented, and reviewed",
            framework_id="nist-csf",
            domain_id="PR",
            risk_level=RiskLevel.MEDIUM,
            control_type=ControlType.TECHNICAL,
            tags={"nist", "cybersecurity", "framework", "logging"}
        ),
        # Detect
        SecurityControl(
            id="DE.AE-1",
            title="Anomalies and Events - Baseline Establishment",
            description="A baseline of network operations and expected data flows for users and systems is established and managed",
            framework_id="nist-csf",
            domain_id="DE",
            risk_level=RiskLevel.MEDIUM,
            control_type=ControlType.TECHNICAL,
            tags={"nist", "cybersecurity", "framework", "detection"}
        ),
        SecurityControl(
            id="DE.CM-1",
            title="Security Continuous Monitoring - System Monitoring",
            description="The network and physical environment is monitored to detect potential cybersecurity events",
            framework_id="nist-csf",
            domain_id="DE",
            risk_level=RiskLevel.HIGH,
            control_type=ControlType.TECHNICAL,
            tags={"nist", "cybersecurity", "framework", "monitoring"}
        ),
        # Respond
        SecurityControl(
            id="RS.RP-1",
            title="Response Planning - Response Plan",
            description="Response plan is executed during or after an incident",
            framework_id="nist-csf",
            domain_id="RS",
            risk_level=RiskLevel.HIGH,
            control_type=ControlType.PROCEDURAL,
            tags={"nist", "cybersecurity", "framework", "incident-response"}
        ),
        SecurityControl(
            id="RS.CO-2",
            title="Communications - Incident Reporting",
            description="Incidents are reported consistent with established criteria",
            framework_id="nist-csf",
            domain_id="RS",
            risk_level=RiskLevel.MEDIUM,
            control_type=ControlType.PROCEDURAL,
            tags={"nist", "cybersecurity", "framework", "incident-response"}
        ),
        # Recover
        SecurityControl(
            id="RC.RP-1",
            title="Recovery Planning - Recovery Plan",
            description="Recovery plan is executed during or after a cybersecurity incident",
            framework_id="nist-csf",
            domain_id="RC",
            risk_level=RiskLevel.HIGH,
            control_type=ControlType.PROCEDURAL,
            tags={"nist", "cybersecurity", "framework", "recovery"}
        ),
        SecurityControl(
            id="RC.IM-1",
            title="Improvements - Lessons Learned",
            description="Recovery plans incorporate lessons learned",
            framework_id="nist-csf",
            domain_id="RC",
            risk_level=RiskLevel.MEDIUM,
            control_type=ControlType.PROCEDURAL,
            tags={"nist", "cybersecurity", "framework", "recovery"}
        )
    ]

    for control in controls:
        framework.add_control(control)

    return framework